docs/docker-loaders.md

```diff
@@ -0,0 +1,76 @@
+2) be able to run multiple versions of loaders at the same run
+3) better fit k8s operation for testing scylla-operator, and have pod (or multiple of them) per stress tool as needed
+4) stop manually maintaining images across multiple regions and cloud vendors
+
+### Images
+
+For most of the tools we are going to create our docker images, under `scylladb/hydra-loaders` and with
+instructions on building each one in `/docker`, we should version them by the tool version if applicable, or by date.
+the only tool we are not going to build ourselves is cassandra-stress, which would be comping from the official docker images
+of scylladb.
+
+### Finding correct cassandra-stress image
+
+Since we are going to be using scylla docker images for loaders, we need a way to find the correct docker image, based on a
+given version or sha of scylla (both for OSS and Enterprise versions), if we'll find that match base on sha would be problematic
+we'll fall back to last available release (smaller than the currently under test version, ex. when testing 5.2-dev, we'll take 5.1.x version)
+
+### Configuration
+
+We should be able to override the default in the test configuration:
+
+```yaml
+stress_image:
+  ycsb: scylladb/hydra-loaders:ycsb-jdk8-20211104
+  cassandra-stress: scylladb/scylla:5.0.0
+  scylla-bench: scylladb/hydra-loaders:scylla-bench-v0.1.8
+```
+
+or via environment variables :
+
+```bash
+export SCT_STRESS_IMAGE='{"ycsb": "scylladb/hydra-loaders:ycsb-jdk8-20211104"}'
+export SCT_STRESS_IMAGE.cassandra-stress="scylladb/scylla:5.0.0"
+export SCT_STRESS_IMAGE.scylla-bench="scylladb/hydra-loaders:scylla-bench-v0.1.8"
+```
+
+for using multiple versions, we should be able to define the image as part of the stress command:
+
+```yaml
+# first would use version 4.6.5, and the 2nd would use the default
+stress_cmd: ["image=scylladb/scylla:4.6.5; cassandra-stress write cl=QUORUM duration=180m ... -rate threads=1000 -pop seq=1..10000000 -log interval=5"
+             "cassandra-stress write cl=QUORUM duration=180m ... -rate threads=1000 -pop seq=1..10000000 -log interval=5"]
+```
+(other option would be to prepend the image, but I think it might be less clear)
+
+this is for cases we might have specific bugs in a nemesis for example that would prevent us from using latest scylla-bench, or latest cassandra-stress
+could be added once we encounter those specific requirements
+
+### Concerns
+
+* Preference tests might have slight different behaviors/timing when working under docker:
+  we should strive to using the host networking, and pin enough CPUs to each stress threads running
+  we will need to run few candidates of the perf tests, with/without those changes, to evaluate.
+
+* for K8S setup we're already using the same image as the image under tests, but switch scylla-bench to docker for example,
+  would mean we'll need to refactor the way create loader pods (since in the current design, we'll end up with a docker running
```

Comment:
Running a docker service inside of a K8S pod is nonsense and
I don't see any reason to describe the situation exactly this way.

It must be something like following:
```
Since we will have separate images for different loader tools and
the need to run more than one during a test run
then we should have the possibility to create loaders of different types.
Having loaders of different types we will be able just to pick up those 
from dedicated python objects queues when needed.
```

Reply:
> Running a docker service inside of a K8S pod is nonsense and I don't see any reason to describe the situation exactly this way.

We are not running docker service inside docker, but using docker client inside docker.
And I'm just explaining the current code/situation here. no that that's the goal or ideal thing to have.

Anyhow we have this issue also if we want to use the docker back-end, but I would be o.k. with docker backend have once solution, and K8S a different one

> 
> It must be something like following:
> 
> ```
> Since we will have separate images for different loader tools and
> the need to run more than one during a test run
> then we should have the possibility to create loaders of different types.
> Having loaders of different types we will be able just to pick up those 
> from dedicated python objects queues when needed.
> ```

Reply:
> @vponomaryov and how we would split the resources for each one ? example, we are using `cassandra-stress` for the main load, but during a nemesis we need `scylla-bench` how much resources we give `scylla-bench`, and how much we give cassandra stress ?
> 
> maybe we should put them in separated pods ? so we can spin them up as needed ?
> 
> in today's VM loader logic, we give each command all CPU (i.e. we don't limit), only if we use `stress_num` parameter of the loader thread, we pin down to specific CPU

Our current implementation for loaders in K8S is to use separate dedicated VM per each loader pod.
And each such pod has dedicated cores.
So, it is not "maybe we should put them in separated pods ?", it must be in separate pods only.
We can spin up loader pod dynamically if needed.
But now we create it during the provisioning stage and run there noop loop to make it be alive.
Then we just run loader commands on demand.

Reply:
@vponomaryov I'm leaving to you, to explain what would be the design for k8s backends, for now it's left as it is.

i.e. for cassandra-stress, since we are using scylla image, we just skip the RemoteDocker part, and let it run on the node/pod. any other stress tool would be triggered as docker as currently implementation enabled it.

I'll need you to look a bit more into it explain the suggestion of how you think it should work, while keeping in mind we don't always know in advance which stress tools are going to be used.